service/matching/taskQueueManager.go

```diff
@@ -630,3 +700,71 @@ func newIOContext() (context.Context, context.CancelFunc) {
 
 	return ctx, cancel
 }
+
+func (c *taskQueueManagerImpl) fetchMetadataFromRootPartitionOnInit(ctx context.Context) {
+	if c.metadataInitialFetchFut.Ready() {
+		return
+	}
+	_, err := c.fetchMetadataFromRootPartition(ctx)
+	c.metadataInitialFetchFut.Set(struct{}{}, err)
+}
+
+// fetchMetadataFromRootPartition fetches metadata from root partition iff this partition is not a root partition.
+// Returns the fetched data, if fetching was necessary and successful.
+func (c *taskQueueManagerImpl) fetchMetadataFromRootPartition(ctx context.Context) (*persistencespb.VersioningData, error) {
+	// Nothing to do if we are the root partition
+	// (for versioning - any later added metadata may need to not abort so early)
+	if c.taskQueueID.IsRoot() {
+		return nil, nil
+	}
+
+	rootTqName := c.taskQueueID.GetRoot()
+	curHash := HashVersioningData(c.db.versioningData)
+	if len(curHash) == 0 {
+		// if we have no data, make sure we send a sigil value, so it's known we desire versioning data
+		curHash = []byte{0}
+	}
+	res, err := c.matchingClient.GetTaskQueueMetadata(ctx, &matchingservice.GetTaskQueueMetadataRequest{
+		NamespaceId:               c.taskQueueID.namespaceID.String(),
+		TaskQueue:                 rootTqName,
+		WantVersioningDataCurhash: curHash,
+	})
+	if err != nil {
+		return nil, err
+	}
+	// If the root partition returns nil here, then that means our data matched, and we don't need to update.
+	// If it's nil because it never existed, then we'd never have any data.
+	// It can't be nil due to removing versions, as that would result in a non-nil container with
+	// nil inner fields.
+	if res.VersioningData != nil {
+		c.db.setVersioningDataForNonRootPartition(res.VersioningData)
+		// Only start the poller if there's actually data to poll for
+		c.metadataPoller.StartIfUnstarted(func() { _, _ = c.fetchMetadataFromRootPartition(context.TODO()) })
+	}
+	return res.VersioningData, nil
+}
+
+func (mp *metadataPoller) StartIfUnstarted(pollFn func()) {
+	mp.launched.Do(func() {
```

Comment:
why take pollFn as an argument? and it only ever starts once? anyway, let's see if this changes at all based on my comment above

Reply:
Original reason is it can't go in when constructed b/c the `taskQueueManagerImpl` doesn't exist yet, but it needs to exist since it's used during polling.

Reply:
it sorta exists, it's in the process of being constructed. if metadataPoller needs a pointer to the tqmi, you can just assign it in the constructor, on a line following the struct literal, to make a cycle. not the cleanest, but I still think that would be better than passing in the anonymous function... the anonymous func in startMetadataPollerIfUnstarted could be a named method on the tqmi